File: gyb_syntax_support/Trivia.py
from .Utils import error
from .kinds import lowercase_first_word  # noqa: I201


class Trivia(object):
    def __init__(self, name, comment, characters=[],
                 swift_characters=[], is_new_line=False, is_comment=False):
        self.name = name
        self.comment = comment
        self.characters = tuple(characters)
        self.lower_name = lowercase_first_word(name)
        if name == "Backslash":
            self.plural_name = "backslashes"
        else:
            self.plural_name = f"{lowercase_first_word(name)}s"
        self.is_new_line = is_new_line
        self.is_comment = is_comment

        # Swift sometimes doesn't support escaped characters like \f or \v;
        # we should allow specifying alternatives explicitly.
        if swift_characters:
            self.swift_characters = tuple(swift_characters)
        else:
            self.swift_characters = characters
        assert len(self.swift_characters) == len(self.characters)

    def characters_len(self):
        return len(self.characters)

    def is_collection(self):
        return self.characters_len() > 0


TRIVIAS = [
    Trivia('Space', 'A space \' \' character.', characters=[' ']),
    Trivia('Tab', 'A tab \'\\t\' character.', characters=['\\t']),
    # Swift don't support vertical tab '\v' so we use the raw unicode
    Trivia('VerticalTab', 'A vertical tab \'\\v\' character.',
           characters=['\\u{b}'], swift_characters=['\\u{2B7F}']),
    # Swift don't support form feed '\f' so we use the raw unicode
    Trivia('Formfeed', 'A form-feed \'f\' character.', characters=['\\u{c}'],
           swift_characters=['\\u{240C}']),
    Trivia('Newline', 'A newline \'\\n\' character.', characters=['\\n'],
           is_new_line=True),
    Trivia('CarriageReturn', 'A newline \'\\r\' character.',
           characters=['\\r'], is_new_line=True),
    Trivia('CarriageReturnLineFeed',
           'A newline consists of contiguous \'\\r\' and \'\\n\' characters.',
           characters=['\\r', '\\n'], is_new_line=True),
    Trivia('LineComment', 'A developer line comment, starting with \'//\'',
           is_comment=True),
    Trivia('BlockComment',
           'A developer block comment, starting with \'/*\' and ending with'
           ' \'*/\'.',
           is_comment=True),
    Trivia('DocLineComment',
           'A documentation line comment, starting with \'///\'.',
           is_comment=True),
    Trivia('DocBlockComment',
           'A documentation block comment, starting with \'/**\' and ending '
           'with \'*/\'.',
           is_comment=True),
    Trivia('Backslash', 'A backslash that is at the end of a line in a multi-line string literal to escape the newline.', characters=['\\\\']),
    Trivia('Pound', 'A \'#\' that is at the end of a line in a multi-line string literal to escape the newline.', characters=['#']),
    Trivia('UnexpectedText', 'Any skipped unexpected text.'),
    Trivia('Shebang', 'A script command, starting with \'#!\'.'),
]

File: CodeGeneration/Sources/SyntaxSupport/Trivia.swift
//===----------------------------------------------------------------------===//
//
// This source file is part of the Swift.org open source project
//
// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
//
//===----------------------------------------------------------------------===//

import SwiftSyntax

public struct TriviaTraits: OptionSet {
  public var rawValue: UInt8

  public init(rawValue: UInt8) {
    self.rawValue = rawValue
  }

  // Indicates this is a whitespace.
  public static var whitespace: Self { .init(rawValue: 1 << 0) }

  // Indicates a newline in Swift source code.
  public static var newline: Self { .init(rawValue: 1 << 1) }

  // Horizontal space.
  public static var spaceOrTab: Self { .init(rawValue: 1 << 2) }

  // Comment in Swift source code.
  public static var comment: Self { .init(rawValue: 1 << 3) }
}

public class Trivia {
  /// The name of the trivia.
  public let name: TokenSyntax

  /// The doc comment describing the trivia.
  public let comment: SwiftSyntax.Trivia

  /// The characters that make up the trivia.
  public let characters: String?

  /// The traits.
  public let traits: TriviaTraits

  /// The name of the trivia in lowercase.
  public var lowerName: TokenSyntax { .identifier(lowercaseFirstWord(name: name.text)) }

  /// The name of the trivia when used as an enum case.
  ///
  /// This is typically the plural form of the `lowerName`, with exceptions for certain trivias.
  public var enumCaseName: TokenSyntax {
    if self.isCollection {
      if lowerName.text == "backslash" {
        return "backslashes"
      } else {
        return .identifier("\(lowerName)s")
      }
    } else {
      return lowerName
    }
  }

  /// Indicates if the trivia is a collection of characters.
  ///
  /// If `true`, the trivia is made up of multiple characters.
  public var isCollection: Bool { characters != nil }

  /// Initializes a new `Trivia` instance.
  ///
  /// - Parameters:
  ///   - name: A name of the trivia.
  ///   - comment: A doc comment describing the trivia.
  ///   - characters: A list of characters that make up the trivia.
  ///   - isComment: Indicates if the trivia represents a comment.
  init(
    name: TokenSyntax,
    comment: SwiftSyntax.Trivia,
    characters: String? = nil,
    traits: TriviaTraits = []
  ) {
    self.name = name
    self.comment = comment
    self.characters = characters
    self.traits = traits
  }
}

public let TRIVIAS: [Trivia] = [
  Trivia(
    name: "Backslash",
    comment: #"A backslash that is at the end of a line in a multi-line string literal to escape the newline."#,
    characters: "\\"
  ),

  Trivia(
    name: "BlockComment",
    comment: #"A developer block comment, starting with '/*' and ending with '*/'."#,
    traits: [.comment]
  ),

  Trivia(
    name: "CarriageReturn",
    comment: #"A newline '\r' character."#,
    characters: "\r",
    traits: [.whitespace, .newline]
  ),

  Trivia(
    name: "CarriageReturnLineFeed",
    comment: #"A newline consists of contiguous '\r' and '\n' characters."#,
    characters: "\r\n",
    traits: [.whitespace, .newline]
  ),

  Trivia(
    name: "DocBlockComment",
    comment: #"A documentation block comment, starting with '/**' and ending with '*/'."#,
    traits: [.comment]
  ),

  Trivia(
    name: "DocLineComment",
    comment: #"A documentation line comment, starting with '///' and excluding the trailing newline."#,
    traits: [.comment]
  ),

  // Swift don't support form feed '\f' so we use the raw unicode
  Trivia(
    name: "Formfeed",
    comment: #"A form-feed 'f' character."#,
    characters: "\u{000C}",
    traits: [.whitespace]
  ),

  Trivia(
    name: "LineComment",
    comment: #"A developer line comment, starting with '//' and excluding the trailing newline."#,
    traits: [.comment]
  ),

  Trivia(
    name: "Newline",
    comment: #"A newline '\n' character."#,
    characters: "\n",
    traits: [.whitespace, .newline]
  ),

  Trivia(
    name: "Pound",
    comment: #"A '#' that is at the end of a line in a multi-line string literal to escape the newline."#,
    characters: "#"
  ),

  Trivia(
    name: "Space",
    comment: #"A space ' ' character."#,
    characters: " ",
    traits: [.whitespace, .spaceOrTab]
  ),

  Trivia(
    name: "Tab",
    comment: #"A tab '\t' character."#,
    characters: "\t",
    traits: [.whitespace, .spaceOrTab]
  ),

  Trivia(
    name: "UnexpectedText",
    comment: #"Any skipped unexpected text."#
  ),

  // Swift don't support vertical tab '\v' so we use the raw unicode
  Trivia(
    name: "VerticalTab",
    comment: #"A vertical tab '\v' character."#,
    characters: "\u{000B}",
    traits: [.whitespace]
  ),
]
